applicationset/controllers/applicationset_controller.go

```diff
@@ -1312,4 +1316,69 @@ func syncApplication(application argov1alpha1.Application, prune bool) (argov1al
 	return application, nil
 }
 
+var ownsHandler = predicate.Funcs{
+	CreateFunc: func(e event.CreateEvent) bool {
+		// if we are the owner and there is a create event, we most likely created it and do not need to
+		// re-reconcile
+		log.Debugln("received create event from owning an application")
+		return false
+	},
+	DeleteFunc: func(e event.DeleteEvent) bool {
+		log.Debugln("received delete event from owning an application")
+		return true
+	},
+	UpdateFunc: func(e event.UpdateEvent) bool {
+		log.Debugln("received update event from owning an application")
+		appOld, isApp := e.ObjectOld.(*argov1alpha1.Application)
+		if !isApp {
+			return false
+		}
+		appNew, isApp := e.ObjectNew.(*argov1alpha1.Application)
+		if !isApp {
+			return false
+		}
+		requeue := shouldRequeueApplicationSet(appOld, appNew)
+		log.Debugf("requeue: %t caused by application %s\n", requeue, appNew.Name)
+		return requeue
+	},
+	GenericFunc: func(e event.GenericEvent) bool {
+		log.Debugln("received generic event from owning an application")
+		return true
+	},
+}
+
+// shouldRequeueApplicationSet determines when we want to requeue an ApplicationSet for reconciling based on an owned
+// application change
+// The applicationset controller owns a subset of the Application CR.
+// We do not need to re-reconcile if parts of the application change outside the applicationset's control.
+// An example being, Application.ApplicationStatus.ReconciledAt which gets updated by the application controller.
+// Additionally, Application.ObjectMeta.ResourceVersion and Application.ObjectMeta.Generation which are set by K8s.
+func shouldRequeueApplicationSet(appOld *argov1alpha1.Application, appNew *argov1alpha1.Application) bool {
+	if appOld == nil || appNew == nil {
+		return false
+	}
+
+	// the applicationset controller owns the application spec, labels, annotations, and finalizers on the applications
+	if !reflect.DeepEqual(appOld.Spec, appNew.Spec) ||
+		!reflect.DeepEqual(appOld.ObjectMeta.GetAnnotations(), appNew.ObjectMeta.GetAnnotations()) ||
+		!reflect.DeepEqual(appOld.ObjectMeta.GetLabels(), appNew.ObjectMeta.GetLabels()) ||
+		!reflect.DeepEqual(appOld.ObjectMeta.GetFinalizers(), appNew.ObjectMeta.GetFinalizers()) {
+		return true
+	}
+
+	// progressive syncs use the application status for updates. if they differ, requeue to trigger the next progression
```

Comment:
Could these comparisons only be included if progressive syncs are enabled? The problem is that they are triggering a lot of unnecessary reconciles of the applicationSet. Here are some examples:
- Every few seconds:
   - Having two apps misconfigured and owning the same resource, ex a deployment. Their `app.Status.Sync.Status` field changes from OutOfSync to Synced every few seconds.
   - Having a statefulset with a volumeClaimTemplate without explicitly saying it is of kind PersistenceVolumeClaim. `app.Status.OperationState.Phase` changes from Running to Succeeded every few seconds.
   
<img width="1028" alt="Screenshot 2023-02-23 at 14 24 59" src="https://user-images.githubusercontent.com/17082952/220920895-27dc2fec-75f4-4888-b34a-757610fa5fdd.png">

- Whenever a deployment occurs:
   - Deploying any resource: As it changes `app.Status.Sync.Status` from OutOfSync to Synced and `OperationState.Phase` from Running to Succeded.
   - Synching an app from the UI: It changes the `app.Status.OperationState.Phase` from Running to Succeeded.

Reply:
I think that we'd ideally work on ways to eliminate excess churn from the state of the Application resources. Like displaying warnings on the Application resource when it appears to be stuck in a "synclooping" state.

Reply:
> Why wouldn't you fix it?

I'm just concerned about the impact application "synclooping" has in the git server. 

From the `application-controller` point of view, it's not very dangerous, as it interacts with git thru the `repo-server`, and it caches `git ls remote` requests, as implemented in this PR: https://github.com/argoproj/argo-cd/pull/6102.

However, the `applicationset-controller` is directly accessing git, without any cache. This translates into argo jamming git whenever one of these sync loops occurs, and bringing our git server down or hitting its rate limiter, until it is detected and manually fixed.

We currently have a monorepo setup, with a git directory generator, deploying 300+ services to 60+ clusters (= 18.000 apps), and have frequently encountered sync loops. That being said, optimizing the applicationset-controller git access it is probably out of the scope of this MR.

Reply:
Oh, I am very aware. We have pretty much the same setup and exact same issues as you. I think [this](https://github.com/argoproj/argo-cd/issues/12407) issue will interest you.

Reply:
> I am of the opinion that changes/syncs from the UI should trigger a reconciliation.

I think that downside only applies to the fields already checked above. Wrapping the `app.Status` checks in progressive-syncs-enabled wouldn't prevent quick reversions of UI changes to Spec fields.

Reply:
I _think_ we are all on the same page and the code reflects that?

Reply:
In `SetupWithManager`, I think we can do `ownsHandler := getOwnsHandler(s.EnableProgressiveSyncs)`. Then modify `shouldRequeueApplicationSet` to accept the boolean and toggle the checks below.

Reply:
added!